File: apps/wallpaper/src/lyrics/lrc.ts
import type { LyricLine } from '@spotify-wallpaper/shared-types';

export interface LrcParseResult {
  offsetMs: number;
  lines: LyricLine[];
}

export interface LyricDisplayState {
  status: 'disabled' | 'missing' | 'before-first-line' | 'active';
  previous: LyricLine | null;
  current: LyricLine | null;
  next: LyricLine | null;
}

export const parseLrc = (input: string): LrcParseResult => {
  let offsetMs = 0;
  const lines: LyricLine[] = [];

  for (const rawLine of input.split(/\r?\n/)) {
    let rest = rawLine.trim();
    const timestamps: number[] = [];

    while (rest.startsWith('[')) {
      const end = rest.indexOf(']');
      if (end < 0) {
        break;
      }

      const tag = rest.slice(1, end);
      rest = rest.slice(end + 1);

      if (tag.startsWith('offset:')) {
        const parsedOffset = Number.parseInt(tag.slice('offset:'.length).trim(), 10);
        offsetMs = Number.isFinite(parsedOffset) ? parsedOffset : 0;
        continue;
      }

      const timestamp = parseTimestamp(tag);
      if (timestamp !== null) {
        timestamps.push(timestamp);
      }
    }

    for (const timestamp of timestamps) {
      lines.push({ timeMs: timestamp + offsetMs, text: rest });
    }
  }

  lines.sort((a, b) => a.timeMs - b.timeMs);
  return { offsetMs, lines };
};

export const lyricDisplayState = (
  lines: LyricLine[],
  progressMs: number,
  enabled: boolean,
  showMissingState: boolean
): LyricDisplayState => {
  if (!enabled) {
    return emptyState('disabled');
  }

  if (lines.length === 0) {
    return emptyState(showMissingState ? 'missing' : 'disabled');
  }

  const currentIndex = findCurrentIndex(lines, progressMs);
  if (currentIndex < 0) {
    return {
      status: 'before-first-line',
      previous: null,
      current: null,
      next: lines[0] ?? null
    };
  }

  return {
    status: 'active',
    previous: lines[currentIndex - 1] ?? null,
    current: lines[currentIndex],
    next: lines[currentIndex + 1] ?? null
  };
};

const parseTimestamp = (tag: string): number | null => {
  const match = /^(\d+):(\d{1,2}(?:\.\d{1,3})?)$/.exec(tag);
  if (!match) {
    return null;
  }

  const minutes = Number.parseInt(match[1], 10);
  const seconds = Number.parseFloat(match[2]);
  if (!Number.isFinite(minutes) || !Number.isFinite(seconds)) {
    return null;
  }

  return minutes * 60_000 + Math.round(seconds * 1000);
};

const findCurrentIndex = (lines: LyricLine[], progressMs: number): number => {
  for (let index = lines.length - 1; index >= 0; index -= 1) {
    if (lines[index].timeMs <= progressMs) {
      return index;
    }
  }

  return -1;
};

const emptyState = (status: LyricDisplayState['status']): LyricDisplayState => ({
  status,
  previous: null,
  current: null,
  next: null
});

File: crates/visual-core/src/lrc.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LyricLine {
    pub time_ms: i64,
    pub text: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LrcParseResult {
    pub offset_ms: i64,
    pub lines: Vec<LyricLine>,
}

pub fn parse_lrc(input: &str) -> LrcParseResult {
    let mut offset_ms = 0;
    let mut lines = Vec::new();

    for raw_line in input.lines() {
        let mut rest = raw_line.trim();
        let mut timestamps = Vec::new();

        while let Some(stripped) = rest.strip_prefix('[') {
            let Some(end) = stripped.find(']') else {
                break;
            };
            let tag = &stripped[..end];
            rest = &stripped[end + 1..];

            if let Some(value) = tag.strip_prefix("offset:") {
                offset_ms = value.trim().parse::<i64>().unwrap_or(0);
                continue;
            }

            if let Some(timestamp) = parse_timestamp(tag) {
                timestamps.push(timestamp);
            }
        }

        for timestamp in timestamps {
            lines.push(LyricLine {
                time_ms: timestamp + offset_ms,
                text: rest.to_string(),
            });
        }
    }

    lines.sort_by_key(|line| line.time_ms);
    LrcParseResult { offset_ms, lines }
}

pub fn current_line(lines: &[LyricLine], progress_ms: i64) -> Option<&LyricLine> {
    lines.iter().rev().find(|line| line.time_ms <= progress_ms)
}

fn parse_timestamp(tag: &str) -> Option<i64> {
    let (minutes, rest) = tag.split_once(':')?;
    let minutes = minutes.parse::<i64>().ok()?;
    let seconds = rest.parse::<f64>().ok()?;
    Some(minutes * 60_000 + (seconds * 1000.0).round() as i64)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_normal_offset_duplicate_and_empty_lines() {
        let parsed = parse_lrc("[offset:100]\n[00:01.00]First\n[00:01.00]Duplicate\n[00:02.50]\n[ar:metadata]");

        assert_eq!(parsed.offset_ms, 100);
        assert_eq!(
            parsed.lines,
            vec![
                LyricLine {
                    time_ms: 1100,
                    text: "First".to_string()
                },
                LyricLine {
                    time_ms: 1100,
                    text: "Duplicate".to_string()
                },
                LyricLine {
                    time_ms: 2600,
                    text: "".to_string()
                }
            ]
        );
    }

    #[test]
    fn finds_current_line_by_progress() {
        let parsed = parse_lrc("[00:01.00]One\n[00:02.00]Two");

        assert_eq!(current_line(&parsed.lines, 500), None);
        assert_eq!(current_line(&parsed.lines, 1500).unwrap().text, "One");
        assert_eq!(current_line(&parsed.lines, 2500).unwrap().text, "Two");
    }
}
